File: scripts/cli/lib/table.js
'use strict';

//NOTE(jimmylee): Table formatting — key-value tables (24ch key) + data tables with column specs.

const { padR, padL, RESET_FG, fgHex, BOLD, COLORS, gradientText } = require('./ansi');

//NOTE(jimmylee): formatRow renders a single table row using column specs for width, alignment,
//NOTE(jimmylee): grow, and status coloring. ColSpec supports optional `gap` (default 1).
function formatRow(vals, colSpec, innerW) {
  const totalGaps = colSpec.reduce((a, c, i) => i > 0 ? a + ((c && c.gap) || 1) : a, 0);
  const baseTotal = colSpec.reduce((a, c) => a + c.width, 0) + totalGaps;
  const extra = Math.max(0, (innerW - 4) - baseTotal);
  const growIdx = colSpec.findIndex((c) => c.grow);
  const widths = colSpec.map((c, i) => c.width + (i === growIdx ? extra : 0));

  let line = '';
  for (let i = 0; i < vals.length; i++) {
    if (i > 0) line += ' '.repeat((colSpec[i] && colSpec[i].gap) || 1);
    let v = vals[i];
    //NOTE(jimmylee): Status coloring uses RESET_FG (not RESET) so the windowBg from card.js
    //NOTE(jimmylee): persists across subsequent columns.
    if (colSpec[i] && colSpec[i].status) {
      if (v === 'ACTIVE' || v === 'OPEN' || v === 'APPROVED') {
        v = `${BOLD}${fgHex(COLORS.success)}${v}${RESET_FG}`;
      } else if (v === 'CLOSED' || v === 'PAID' || v === 'SUSPENDED') {
        v = `${fgHex(COLORS.neutral)}${v}${RESET_FG}`;
      }
    }
    const align = colSpec[i] && colSpec[i].align;
    line += align === 'right' ? padL(v, widths[i]) : padR(v, widths[i]);
  }
  return line;
}

function kvTable(pairs) {
  const lines = [];
  for (const [k, v] of pairs) {
    lines.push(`${padR(k, 24)}${v}`);
  }
  return lines;
}

//NOTE(jimmylee): Key-value table with gradient on value column (matches .gradientText td:last-child).
function kvTableGradient(pairs) {
  const lines = [];
  for (const [k, v] of pairs) {
    lines.push(`${padR(k, 24)}${gradientText(v, COLORS.gradientStart, COLORS.gradientEnd)}`);
  }
  return lines;
}

module.exports = { formatRow, kvTable, kvTableGradient };

File: scripts/python/sacred_cli/table.py
"""Table formatting. Mirrors scripts/cli/lib/table.js."""

from .ansi import BOLD, COLORS, RESET_FG, fg_hex, gradient_text, pad_l, pad_r


def format_row(vals, col_spec, inner_w: int) -> str:
    total_gaps = sum((c.get("gap", 1) if i > 0 else 0) for i, c in enumerate(col_spec))
    base_total = sum(c["width"] for c in col_spec) + total_gaps
    extra = max(0, (inner_w - 4) - base_total)
    grow_idx = next((i for i, c in enumerate(col_spec) if c.get("grow")), -1)
    widths = [c["width"] + (extra if i == grow_idx else 0) for i, c in enumerate(col_spec)]

    line = ""
    for i, v in enumerate(vals):
        if i > 0:
            line += " " * col_spec[i].get("gap", 1)
        if col_spec[i].get("status"):
            if v in ("ACTIVE", "OPEN", "APPROVED"):
                v = f"{BOLD}{fg_hex(COLORS['success'])}{v}{RESET_FG}"
            elif v in ("CLOSED", "PAID", "SUSPENDED"):
                v = f"{fg_hex(COLORS['neutral'])}{v}{RESET_FG}"
        align = col_spec[i].get("align")
        line += pad_l(v, widths[i]) if align == "right" else pad_r(v, widths[i])
    return line


def kv_table(pairs):
    return [f"{pad_r(k, 24)}{v}" for k, v in pairs]


def kv_table_gradient(pairs):
    return [
        f"{pad_r(k, 24)}{gradient_text(v, COLORS['gradientStart'], COLORS['gradientEnd'])}"
        for k, v in pairs
    ]
